aesd-char-driver/main.c

```diff
@@ -17,54 +17,130 @@
 #include <linux/types.h>
 #include <linux/cdev.h>
 #include <linux/fs.h> // file_operations
+#include <linux/slab.h>
+#include <asm/bug.h>
 #include "aesdchar.h"
 int aesd_major =   0; // use dynamic major
 int aesd_minor =   0;
 
-MODULE_AUTHOR("Your Name Here"); /** TODO: fill in your name **/
+#define COMBUF_INITCAP 1024
+
+#define LOCK_DEV(d) do {} while(down_interruptible(&d.sem))
+
+MODULE_AUTHOR("George Hodgkins");
 MODULE_LICENSE("Dual BSD/GPL");
 
-struct aesd_dev aesd_device;
+struct aesd_dev the_dev;
+struct file_operations aesd_fops;
 
 int aesd_open(struct inode *inode, struct file *filp)
 {
 	PDEBUG("open");
-	/**
-	 * TODO: handle open
-	 */
+	if (filp->f_op != &aesd_fops) {
+	   PDEBUG("aesdchar: f_ops seems wrong: is %p, should be %p", filp->f_op, &aesd_fops);	
+	}
 	return 0;
 }
 
 int aesd_release(struct inode *inode, struct file *filp)
 {
 	PDEBUG("release");
-	/**
-	 * TODO: handle release
-	 */
 	return 0;
 }
 
 ssize_t aesd_read(struct file *filp, char __user *buf, size_t count,
                 loff_t *f_pos)
 {
-	ssize_t retval = 0;
-	PDEBUG("read %zu bytes with offset %lld",count,*f_pos);
-	/**
-	 * TODO: handle read
-	 */
-	return retval;
+	if (!access_ok(buf, count)) return -EFAULT;
+	LOCK_DEV(the_dev);
```

Comment:
@georgehodgkins consider using `private_data` to access your dev structure instead.  Not required for this simple assignment but would allow your code to work with multiple dev endpoints.